src/main/java/net/minecraftforge/registries/ForgeRegistry.java

```diff
@@ -234,26 +240,164 @@ public ResourceLocation getKey(V value)
+    {
+        if (!this.hasWrapper)
+            throw new IllegalStateException("Cannot query holder helper for non-wrapped forge registry!");
+
+        return this.defaultKey != null
+                ? this.getSlaveMap(NamespacedDefaultedWrapper.Factory.ID, NamespacedDefaultedWrapper.class).getHolderHelper()
+                : this.getSlaveMap(NamespacedWrapper.Factory.ID, NamespacedWrapper.class).getHolderHelper();
+    }
+
+    @NotNull
+    @Override
+    public Optional<Holder<V>> getHolder(ResourceKey<V> key)
+    {
+        return getHolderHelper().getHolder(key);
+    }
+
+    @Override
+    public @NotNull Holder<V> getHolderOrThrow(ResourceKey<V> key)
+    {
+        return this.getHolder(key).orElseThrow(() -> new IllegalStateException("Missing key in " + this.key + ": " + key));
+    }
+
+    @NotNull
+    @Override
+    public Optional<Holder<V>> getHolder(ResourceLocation location)
+    {
+        return getHolderHelper().getHolder(location);
+    }
+
+    @NotNull
+    @Override
+    public Holder<V> getHolderOrThrow(ResourceLocation location)
+    {
+        return this.getHolder(location).orElseThrow(() -> new IllegalStateException("Missing key in " + this.key + ": " + location));
+    }
+
+    @NotNull
+    @Override
+    public Optional<Holder<V>> getHolder(V value)
+    {
+        return getHolderHelper().getHolder(value);
+    }
+
+    @Override
+    public @NotNull Holder<V> getHolderOrThrow(V value)
+    {
+        return this.getHolder(value).orElseThrow(() -> new IllegalStateException("Missing value in " + this.key + ": " + value));
+    }
+
+    @NotNull
+    @Override
+    public HolderSet.Named<V> getOrCreateTag(TagKey<V> name)
+    {
+        return getHolderHelper().getOrCreateTag(name);
+    }
+
+    @NotNull
+    @Override
+    public Optional<HolderSet.Named<V>> getTag(TagKey<V> name)
+    {
+        return getHolderHelper().getTag(name);
+    }
+
+    @NotNull
+    @Override
+    public Iterable<Holder<V>> getTagOrEmpty(TagKey<V> name)
+    {
+        return DataFixUtils.orElse(this.getTag(name), List.of());
+    }
+
+    @NotNull
+    @Override
+    public boolean isKnownTagName(TagKey<V> name)
+    {
+        return getHolderHelper().isKnownTagName(name);
+    }
+
+    @NotNull
+    @Override
+    public Stream<Pair<TagKey<V>, HolderSet.Named<V>>> getTags()
+    {
+        return getHolderHelper().getTags();
+    }
+
+    @NotNull
+    @Override
+    public Stream<TagKey<V>> getTagNames()
+    {
+        return getHolderHelper().getTagNames();
+    }
+
+    @NotNull
+    @Override
+    public TagKey<V> createTagKey(ResourceLocation location)
+    {
+        return TagKey.create(this.getRegistryKey(), location);
+    }
+
+    @NotNull
+    @Override
+    public TagKey<V> createOptionalTagKey(ResourceLocation location, @NotNull Set<Supplier<V>> defaults)
+    {
+        TagKey<V> tagKey = createTagKey(location);
+
+        getHolderHelper().addOptionalTag(tagKey, defaults);
+
+        return tagKey;
+    }
```

Comment:
This probably should be advertised when this PR is merged as the replacement for the old `createOptional` with defaults.